### PR TITLE
llvmPackages_{12,13,14,15,16,17,18,19,git}: commonify patches

### PR DESCRIPTION
**Pull request by RossComputerGuy**
## Description of changes

Reduce the number of patches we have by finding duplicates. Modify `getVersionFile` in `common/default.nix` to branch to min/max version patches so we only have to keep 1 copy of a specific version of a patch.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by RossComputerGuy** [comment]
Swift build failures can be ignored as they are happening on Hydra as well. It looks like it at least passes the `patchPhase`.

File: pkgs/development/compilers/llvm/common/default.nix [hunk 1]
```diff
@@ -74,7 +74,98 @@ let
       p:
       builtins.path {
         name = builtins.baseNameOf p;
-        path = "${metadata.versionDir}/${p}";
+        path =
```
Comment (alyssais) [inline pkgs/development/compilers/llvm/common/default.nix:77]: This big nested conditional is not great — I think maybe it indicates that getVersionFile is not the right abstraction.  Maybe it'd be better to just have a list of patches for each major LLVM version?

**Review by alyssais — COMMENTED** [review]
This is fantastic work — look at all those deletions!  Thank you!

I'm just wondering if there's a better way to express it…

File: pkgs/development/compilers/llvm/common/default.nix (hunk 1 above)
Comment (RossComputerGuy) [reply pkgs/development/compilers/llvm/common/default.nix:77]: Maybe but there's unconditional patches and some patches are the same between versions while having different siblings. I was thinking of doing an attr set and have it work like a table but I found it to be simpler to do it this way.

I was thinking of doing a PR to clean up the common default.nix so it wouldn't be as sloppy. Probably in that one, I could have an attr set of versions and values which could be a string or null which then matches into figuring out where that patch will be.

Comment (alyssais) [reply pkgs/development/compilers/llvm/common/default.nix:77]: > I was thinking of doing an attr set and have it work like a table

This is what I was thinking of.  Was the problem just that you'd have to repeat the patch of a patch multiple times to say that different versions have the same patch?  If so, I think that would be easier to understand overall.

Comment (RossComputerGuy) [reply pkgs/development/compilers/llvm/common/default.nix:77]: Maybe, I'm not sure.

**Review by alyssais — APPROVED** [review]
This is still very complicated, but it's better, and maybe there's just not a better way (apart from upstreaming the patches, of course).

Collapsing the attrsets into single lines (if that's allowed by formatting rules) might help, by making it a bit denser.

**Comment by RossComputerGuy** [comment]
> Collapsing the attrsets into single lines (if that's allowed by formatting rules) might help, by making it a bit denser.

Yeah, I originally did that but the formatter changed it.

**Review by Ericson2314 — APPROVED** [review]

**Comment by RossComputerGuy** [comment]
Before we merge, I want to be able to reproduce and fix that CI error. However, I tried and couldn't get it to fail locally.

**Comment by RossComputerGuy** [comment]
Well, by-name passed so we should be gtg now.

File: pkgs/development/compilers/llvm/common/default.nix [hunk 2]
```diff
@@ -74,7 +74,222 @@ let
...
+                }
+              ];
+              "compiler-rt/armv7l.patch" = [
+                {
+                  before = "15";
+                  after = "13";
+                  path = ../13;
+                }
+              ];
+              "compiler-rt/gnu-install-dirs.patch" = [
```
Comment (tomberek) [inline pkgs/development/compilers/llvm/common/default.nix:196]: possible error here?

**Pull request merged**